drivers/sensor/lis2mdl/lis2mdl.c

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright (c) 2018 STMicroelectronics
+ *
+ * LIS2MDL mag driver
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <misc/__assert.h>
+#include <misc/byteorder.h>
+#include <sensor.h>
+#include <string.h>
+
+#include "lis2mdl.h"
+
+struct lis2mdl_data lis2mdl_device_data;
+
+int lis2mdl_raw_read8(struct lis2mdl_data *lis2mdl, u16_t dev_addr,
+		       u8_t reg_addr, u8_t *value)
+{
+	return i2c_reg_read_byte(lis2mdl->i2c, dev_addr, reg_addr, value);
+}
+
+int lis2mdl_raw_write8(struct lis2mdl_data *lis2mdl, u16_t dev_addr,
+			u8_t reg_addr, u8_t value)
+{
+	return i2c_reg_write_byte(lis2mdl->i2c, dev_addr, reg_addr, value);
+}
+
+int lis2mdl_raw_read_n(struct lis2mdl_data *lis2mdl, u16_t dev_addr,
+			u8_t start_addr, u8_t *buf,
+			u8_t num_bytes)
+{
+	/* set i2c auto increment */
+	return i2c_burst_read(lis2mdl->i2c, dev_addr,
+			      start_addr | LIS2MDL_AUTO_INCREMENT,
+			      buf, num_bytes);
+}
+
+int lis2mdl_raw_write_n(struct lis2mdl_data *lis2mdl, u16_t dev_addr,
+			 u8_t start_addr, u8_t *buf,
+			 u8_t num_bytes)
+{
+	/* set i2c auto increment */
+	return i2c_burst_write(lis2mdl->i2c, dev_addr,
+			       start_addr | LIS2MDL_AUTO_INCREMENT,
+			       buf, num_bytes);
+}
+
+int lis2mdl_write_data_with_mask(struct lis2mdl_data *lis2mdl,
+				  u8_t reg_addr,
+				  u8_t mask, u8_t data)
+{
+	return i2c_reg_update_byte(lis2mdl->i2c, CONFIG_LIS2MDL_I2C_ADDR,
+				   reg_addr, mask, data);
+}
+
+static int lis2mdl_set_odr(struct device *dev, u16_t odr)
+{
+	u8_t reg_val;
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+
+	/* check if power off */
+	if (odr == 0) {
+		/* power off mag */
+		return lis2mdl_write_data_with_mask(lis2mdl, LIS2MDL_CFG_REG_A,
+						    LIS2MDL_MAG_MODE_MASK,
+						    LIS2MDL_MD_IDLE1_MODE);
+	}
+
+	reg_val = LIS2MDL_ODR_TO_REG(odr);
+	if (reg_val > LIS2MDL_ODR100_HZ) {
+		reg_val = LIS2MDL_ODR100_HZ;
+	} else if (reg_val < LIS2MDL_ODR10_HZ) {
+		reg_val = LIS2MDL_ODR10_HZ;
+	}
+
+	return lis2mdl_write_data_with_mask(lis2mdl, LIS2MDL_CFG_REG_A,
+					    LIS2MDL_ODR_MASK, reg_val);
+}
+
+static int lis2mdl_set_hard_iron(struct device *dev, enum sensor_channel chan,
+				  const struct sensor_value *val)
+{
+	struct lis2mdl_data *lis2mdl = dev->driver_data;
+	u8_t regs = LIS2MDL_OFFSET_X_REG_L;
+	u8_t i;
+	s16_t offs;
+	int ret;
+
+	for (i = 0; i < LIS2MDL_NUM_AXIS; i++) {
+		offs = val->val1;
+		ret = lis2mdl_raw_write_n(lis2mdl, CONFIG_LIS2MDL_I2C_ADDR,
+					  regs, (u8_t *)&offs,
```

Comment:
You assume the sensor and MCU have the same endianness. Use a byte order macro in `include/misc/byteorder.h`